samza-core/src/main/scala/org/apache/samza/system/StreamMetadataCache.scala

```diff
@@ -94,17 +112,120 @@ class StreamMetadataCache (
+    * requestedSystemStreamPartition.
+    */
+  private def fetchNewestOffsetWithPrefetching(requestedSystemStreamPartition: SystemStreamPartition,
+                                               maxRetries: Integer,
+                                               now: Long): String = {
+    val systemToFetch = requestedSystemStreamPartition.getSystem
+    /*
+     * If a systemStreamPartition is in the newestOffsetCache, then we have fetched it before, so include it in
+     * the metadata fetch request now if it is stale. Also make sure it has the same system, since we will only use the
+     * admin for the system of the requestedSystemStreamPartition.
+     */
+    var systemStreamPartitionsToFetch = newestOffsetCache
+      .filter({case (ssp, cacheEntry) => systemToFetch.equals(ssp.getSystem) && isStale(now, cacheEntry.lastRefreshMs)})
+      .keySet.toSet
+    if (!systemStreamPartitionsToFetch.contains(requestedSystemStreamPartition)) {
+      systemStreamPartitionsToFetch += requestedSystemStreamPartition
+    }
+    val newestOffsets: Map[SystemStreamPartition, String] = systemAdmins.getSystemAdmin(systemToFetch) match {
+      case extendedSystemAdmin: ExtendedSystemAdmin =>
+        /*
+         * ExtendedSystemAdmin has a way to fetch the newest offsets for specific SSPs instead of fetching the
+         * newest and oldest offsets for all SSPs. Use it if we can.
+         */
+        extendedSystemAdmin.getNewestOffsets(systemStreamPartitionsToFetch.asJava, maxRetries).asScala.toMap
+      case systemAdmin => fetchNewestOffsetsForSystemAdmin(systemStreamPartitionsToFetch, systemAdmin)
+    }
+    addNewestOffsetsToCache(systemStreamPartitionsToFetch, newestOffsets, now)
+    newestOffsets.get(requestedSystemStreamPartition) match {
+      case Some(offset) => offset
+      case None => null
     }
   }
+
+  /**
+    * Add the newest offsets data to the cache.
+    * The systemStreamPartitionsToFetch are needed because newestOffsets might not have an entry for all SSPs.
+    */
+  def addNewestOffsetsToCache(systemStreamPartitionsToFetch: Set[SystemStreamPartition],
+                              newestOffsets: Map[SystemStreamPartition, String],
+                              time: Long): Unit = {
+    systemStreamPartitionsToFetch.foreach(systemStreamPartition => {
+      // want to put an entry in regardless of if we got data from newestOffsets
+      newestOffsetCache.update(systemStreamPartition, CacheEntry(newestOffsets.get(systemStreamPartition), time))
+    })
+  }
+
+  def fetchNewestOffsetsForSystemAdmin(systemStreamPartitionsToFetch: Set[SystemStreamPartition],
+                                       systemAdmin: SystemAdmin): Map[SystemStreamPartition, String] = {
+    val streamsToFetch = systemStreamPartitionsToFetch
+      .map(systemStreamPartitionToFetch => systemStreamPartitionToFetch.getStream)
+    val streamToMetadata = systemAdmin.getSystemStreamMetadata(streamsToFetch.asJava).asScala.toMap
```

Comment:
Maybe update the full metadata cache as well if we already have the information?